File: python/redfish_sdk/services/ras_service.py
# SPDX-License-Identifier: MIT
# Copyright (c) Microsoft Corporation. All rights reserved.

"""
redfish_sdk/services/ras_service.py

RasService handle — RAS API endpoint discovery, CPER event subscription,
large-CPER retrieval via AdditionalDataUri, and CPAD submission.

Terminology (OCP Fault Management Infrastructure / UEFI RAS API spec):
  CPER  — Common Platform Error Record (UEFI 2.9A)
  CPAD  — Common Platform Action Descriptor (OCP RAS API v1.0)
  CreatorID   — GUID identifying the vendor analyzer for a set of CPERs.
  PartitionID — BMC-assigned ID that uniquely identifies a silicon endpoint;
                used to route CPADs back to the right component.
  FRU ID/Text — Field-Replaceable Unit identifier / human-readable location.
"""

from __future__ import annotations

import asyncio
import base64
import json
import logging
from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING

from redfish_sdk.protocol.response import RedfishResponse, build_response
from redfish_sdk.transport.auth import AuthManager

if TYPE_CHECKING:
    from redfish_sdk.transport.http_client import HttpClient
    from redfish_sdk.models.redfish_types import AuthState

logger = logging.getLogger(__name__)

# Default RAS service root path.  Update to the standardised OCP URI once
# the RAS API Redfish registry is finalised.
_DEFAULT_SERVICE_PATH = "/redfish/v1/RasService"

# Registry prefixes used when subscribing to CPER events.  Populate this list
# with the finalised OCP RAS registry prefix (e.g. "RASEvent") before use.
RAS_REGISTRY_PREFIXES: list[str] = []


# ---------------------------------------------------------------------------
# CPER severity / queue types
# ---------------------------------------------------------------------------

class CperSeverity(Enum):
    """
    Maps to the five RAS API CPER queues defined in the OCP RAS API v1.0 spec.

    Queue         Description
    ─────────────────────────────────────────────────────────────────────────
    PlatformEvent Informational CPERs / Platform Action Events (not errors)
    Informational Deferred errors including poison generation
    Corrected     Hardware-corrected errors
    Recoverable   Errors where the OS may survive (poison consumption, PCIe)
    Fatal         Errors that crash the OS / generate hardware crashdumps
    """
    PLATFORM_EVENT = "PlatformEvent"
    INFORMATIONAL  = "Informational"
    CORRECTED      = "Corrected"
    RECOVERABLE    = "Recoverable"
    FATAL          = "Fatal"

    @classmethod
    def from_message_id(cls, message_id: str) -> "CperSeverity | None":
        """Infer severity from a Redfish MessageId string (case-insensitive)."""
        lower = message_id.lower()
        for sev in cls:
            if sev.value.lower() in lower:
                return sev
        return None


# ---------------------------------------------------------------------------
# Data models
# ---------------------------------------------------------------------------

@dataclass
class RasEndpoint:
    """
    A silicon RAS API endpoint discovered by the BMC and exposed via Redfish.

    The BMC performs MCTP/PLDM enumeration and populates these fields from
    the RAS API discovery exchange with each endpoint.
    """
    endpoint_id:      str
    creator_id:       str               # GUID — identifies the vendor analyzer
    fru_id:           str               # GUID — unique FRU instance identifier
    partition_id:     str               # BMC-assigned routing ID for CPADs
    supported_queues: list[str] = field(default_factory=list)
    uri:              str = ""          # Redfish URI for this endpoint resource
    raw:              dict = field(default_factory=dict)


@dataclass
class CperEvent:
    """
    A CPER-carrying Redfish event delivered by the BMC.

    For small CPERs the ``cper_data`` field is populated directly with the
    decoded binary payload.  For large CPERs the BMC sets ``additional_data_uri``
    instead; call ``RasServiceHandle.fetch_cper_data()`` to retrieve the blob.
    """
    event_id:            str
    message_id:          str
    severity:            CperSeverity | None
    timestamp:           str
    origin_of_condition: str | None = None
    cper_data:           bytes | None = None    # inline CPER bytes (base64-decoded)
    additional_data_uri: str | None = None      # URI for large-CPER retrieval
    raw:                 dict = field(default_factory=dict)

    @classmethod
    def from_event_record(cls, record: dict) -> "CperEvent":
        """Parse a single EventRecord dict from a Redfish EventMessage payload."""
        message_id = record.get("MessageId", "")
        severity   = CperSeverity.from_message_id(message_id)

        # Inline CPER: may be base64-encoded in AdditionalData or a vendor
        # extension field such as Oem.CperData.
        cper_data: bytes | None = None
        raw_cper = (
            record.get("AdditionalData")
            or record.get("Oem", {}).get("CperData")
        )
        if isinstance(raw_cper, str):
            try:
                cper_data = base64.b64decode(raw_cper)
            except Exception:
                pass

        return cls(
            event_id            = record.get("EventId", ""),
            message_id          = message_id,
            severity            = severity,
            timestamp           = record.get("EventTimestamp", ""),
            origin_of_condition = record.get("OriginOfCondition", {}).get("@odata.id"),
            cper_data           = cper_data,
            additional_data_uri = record.get("AdditionalDataURI"),
            raw                 = record,
        )


@dataclass
class CpadRecord:
    """
    A Common Platform Action Descriptor (CPAD) to submit to the BMC.

    The BMC uses ``partition_id`` to route the action to the correct silicon
    endpoint.  ``platform_id`` must match the value the BMC advertises in its
    own CPERs so the fleet infrastructure can direct the CPAD here.

    ``payload`` is the raw binary CPAD blob as defined in the OCP RAS API spec.
    It is transmitted base64-encoded in the JSON body.
    """
    platform_id:  str               # identifies this BMC within the fleet
    partition_id: str               # identifies the target silicon endpoint
    creator_id:   str               # identifies the issuing analyzer
    payload:      bytes             # binary CPAD blob
    fru_id:       str = ""
    fru_text:     str = ""          # physical location label (e.g. silkscreen)


# ---------------------------------------------------------------------------
# Service handle
# ---------------------------------------------------------------------------

class RasServiceHandle:
    """
    Redfish client handle for the RAS API service.

    Provides:
      - Discovery of BMC-exposed RAS API endpoints (silicon components).
      - CPER event subscription (wraps EventService with RAS message IDs).
      - Large-CPER retrieval via AdditionalDataUri.
      - CPAD submission (PUT to the BMC's CPAD URI).
      - Static helper to parse CperEvent objects from a push-mode payload.
    """

    def __init__(
        self,
        http: "HttpClient",
        auth_state: "AuthState",
        discovery_map: dict[str, str],
    ) -> None:
        self._http = http
        self._auth_state = auth_state
        self._discovery_map = discovery_map

    @property
    def _service_uri(self) -> str:
        return self._discovery_map.get("RasService", _DEFAULT_SERVICE_PATH)

    # ------------------------------------------------------------------
    # Endpoint discovery
    # ------------------------------------------------------------------

    async def discover_endpoints_async(self) -> list[RasEndpoint]:
        """
        Query the BMC for all discovered RAS API endpoints.

        The BMC populates this collection during MCTP/PLDM enumeration.
        Each member includes Creator ID, FRU ID, Partition ID, and the
        supported CPER queues confirmed during RAS API discovery.
        """
        headers = AuthManager.attach_auth(self._auth_state, {})
        raw = await self._http.request_async("GET", self._service_uri, headers=headers)
        resp = build_response(raw.status_code, raw.headers, raw.body_json, raw.body_text)
        if not resp.success:
            logger.warning("RAS endpoint discovery failed: HTTP %s", resp.status_code)
            return []

        endpoints: list[RasEndpoint] = []
        for member in resp.body.get("Members", []):
            uri = member.get("@odata.id", "")
            if not uri:
                continue
            dr = await self._http.request_async("GET", uri, headers=headers)
            detail = build_response(dr.status_code, dr.headers, dr.body_json, dr.body_text)
            if detail.success:
                d = detail.body
                endpoints.append(RasEndpoint(
                    endpoint_id      = d.get("Id", ""),
                    creator_id       = d.get("CreatorId", ""),
                    fru_id           = d.get("FruId", ""),
                    partition_id     = d.get("PartitionId", ""),
                    supported_queues = d.get("SupportedQueues", []),
                    uri              = uri,
                    raw              = d,
                ))

        logger.debug("RAS: discovered %d endpoint(s)", len(endpoints))
        return endpoints

    def discover_endpoints(self) -> list[RasEndpoint]:
        return asyncio.run(self.discover_endpoints_async())

    # ------------------------------------------------------------------
    # CPER event subscription
    # ------------------------------------------------------------------

    async def subscribe_cper_events_async(
        self,
        destination:       str,
        registry_prefixes: list[str] | None = None,
        message_ids:       list[str] | None = None,
        context:           str = "RAS-CPER",
        event_format_type: str = "Event",
    ) -> RedfishResponse:
        """
        Subscribe to CPER-carrying Redfish events from this BMC.

        ``registry_prefixes`` and ``message_ids`` narrow the subscription to
        RAS-specific events.  If both are omitted, ``RAS_REGISTRY_PREFIXES`` is
        used (update that list once the OCP registry name is finalised; sending
        an empty subscription receives all events from the BMC).
        """
        effective_prefixes = (
            registry_prefixes if registry_prefixes is not None
            else list(RAS_REGISTRY_PREFIXES)
        )

        body: dict = {
            "Destination":     destination,
            "Protocol":        "Redfish",
            "SubscriptionType": "RedfishEvent",
            "Context":         context,
            "EventFormatType": event_format_type,
        }
        if effective_prefixes:
            body["RegistryPrefixes"] = effective_prefixes
        if message_ids:
            body["MessageIds"] = message_ids

        subs_uri = self._discovery_map.get("EventService", "/redfish/v1/EventService")
        subs_uri = f"{subs_uri}/Subscriptions"

        headers = AuthManager.attach_auth(self._auth_state, {})
        raw = await self._http.request_async("POST", subs_uri, headers=headers, body=body)
        return build_response(raw.status_code, raw.headers, raw.body_json, raw.body_text)

    def subscribe_cper_events(self, destination: str, **kwargs) -> RedfishResponse:
        return asyncio.run(self.subscribe_cper_events_async(destination, **kwargs))

    # ------------------------------------------------------------------
    # Large-CPER retrieval
    # ------------------------------------------------------------------

    async def fetch_cper_data_async(self, additional_data_uri: str) -> bytes:
        """
        Retrieve a CPER payload from the BMC via the ``AdditionalDataUri``
        carried in a Redfish event.  Returns the raw bytes of the CPER blob.

        The BMC may encode the binary payload as base64 inside a JSON field
        (``CperData``, ``Data``, or ``AdditionalData``) or serve it as raw
        ``application/octet-stream``.  Both cases are handled.
        """
        headers = AuthManager.attach_auth(self._auth_state, {})
        raw = await self._http.request_async("GET", additional_data_uri, headers=headers)
        if raw.status_code < 200 or raw.status_code >= 300:
            raise RuntimeError(
                f"Failed to fetch CPER data from {additional_data_uri!r}: "
                f"HTTP {raw.status_code}"
            )

        # JSON response — try known base64 fields in priority order
        if raw.body_json and isinstance(raw.body_json, dict):
            for key in ("CperData", "Data", "AdditionalData"):
                if key in raw.body_json:
                    try:
                        return base64.b64decode(raw.body_json[key])
                    except Exception:
                        pass
            # Fallback: return raw JSON serialised as bytes
            return json.dumps(raw.body_json).encode()

        # Raw binary / text fallback
        return (raw.body_text or "").encode()

    def fetch_cper_data(self, additional_data_uri: str) -> bytes:
        return asyncio.run(self.fetch_cper_data_async(additional_data_uri))

    # ------------------------------------------------------------------
    # CPAD submission
    # ------------------------------------------------------------------

    async def submit_cpad_async(self, cpad_uri: str, cpad: CpadRecord) -> RedfishResponse:
        """
        Submit a CPAD to the BMC via HTTP PUT.

        The BMC acknowledges receipt in the response (does not imply the action
        has been taken).  Action completion is signalled asynchronously via a
        Platform Action Event CPER on the Event queue.

        ``cpad_uri`` is the BMC's CPAD endpoint URI.  Discovery of this URI is
        not yet standardised; obtain it from the service root or BMC documentation.
        """
        body = {
            "PlatformId":  cpad.platform_id,
            "PartitionId": cpad.partition_id,
            "CreatorId":   cpad.creator_id,
            "FruId":       cpad.fru_id,
            "FruText":     cpad.fru_text,
            "Payload":     base64.b64encode(cpad.payload).decode(),
        }
        headers = AuthManager.attach_auth(self._auth_state, {})
        raw = await self._http.request_async("PUT", cpad_uri, headers=headers, body=body)
        return build_response(raw.status_code, raw.headers, raw.body_json, raw.body_text)

    def submit_cpad(self, cpad_uri: str, cpad: CpadRecord) -> RedfishResponse:
        return asyncio.run(self.submit_cpad_async(cpad_uri, cpad))

    # ------------------------------------------------------------------
    # Parse CPER events from a push-mode Redfish event payload
    # ------------------------------------------------------------------

    @staticmethod
    def parse_cper_events(event_payload: dict) -> list[CperEvent]:
        """
        Extract ``CperEvent`` objects from a Redfish EventMessage payload dict
        (the JSON body POSTed to your event listener by the BMC).

        Filters to records that are identifiable as RAS/CPER events — those
        where a severity can be inferred or the MessageId contains "cper"/"ras".
        Non-RAS events in the same payload are silently ignored.
        """
        events: list[CperEvent] = []
        for record in event_payload.get("Events", []):
            ev = CperEvent.from_event_record(record)
            mid_lower = ev.message_id.lower()
            if ev.severity is not None or "cper" in mid_lower or "ras" in mid_lower:
                events.append(ev)
        return events

File: rust/redfish-sdk/src/services/ras_service.rs
// SPDX-License-Identifier: MIT
// Copyright (c) Microsoft Corporation. All rights reserved.

//! RAS API service handle — endpoint discovery, CPER event subscription,
//! large-CPER retrieval via AdditionalDataUri, and CPAD submission.
//!
//! Terminology:
//! - **CPER**  — Common Platform Error Record (UEFI 2.9A)
//! - **CPAD**  — Common Platform Action Descriptor (OCP RAS API v1.0)
//! - **CreatorID**   — GUID identifying the vendor analyzer for a CPER stream.
//! - **PartitionID** — BMC-assigned routing ID used to direct CPADs to the
//!                     correct silicon endpoint.

use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use serde_json::json;

use crate::context::ClientContext;
use crate::errors::RedfishError;
use crate::protocol::response::RedfishResponse;

// ---------------------------------------------------------------------------
// CPER severity / queue types
// ---------------------------------------------------------------------------

/// Maps to the five RAS API CPER queues defined in the OCP RAS API v1.0 spec.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CperSeverity {
    /// Informational / Platform Action Events (not error events)
    PlatformEvent,
    /// Deferred errors including poison generation
    Informational,
    /// Hardware-corrected errors
    Corrected,
    /// OS-survivable errors (poison consumption, PCIe device failure)
    Recoverable,
    /// OS-crashing errors / hardware crashdumps
    Fatal,
}

impl CperSeverity {
    /// Infer severity from a Redfish MessageId string (case-insensitive).
    pub fn from_message_id(message_id: &str) -> Option<Self> {
        let lower = message_id.to_lowercase();
        if lower.contains("platformevent") { return Some(Self::PlatformEvent); }
        if lower.contains("informational")  { return Some(Self::Informational); }
        if lower.contains("corrected")      { return Some(Self::Corrected); }
        if lower.contains("recoverable")    { return Some(Self::Recoverable); }
        if lower.contains("fatal")          { return Some(Self::Fatal); }
        None
    }
}

// ---------------------------------------------------------------------------
// Data models
// ---------------------------------------------------------------------------

/// A silicon RAS API endpoint discovered by the BMC and exposed via Redfish.
#[derive(Debug, Clone)]
pub struct RasEndpoint {
    pub endpoint_id:      String,
    pub creator_id:       String,           // GUID — identifies the vendor analyzer
    pub fru_id:           String,           // GUID — unique FRU instance identifier
    pub partition_id:     String,           // BMC-assigned routing ID for CPADs
    pub supported_queues: Vec<String>,
    pub uri:              String,
    pub raw:              serde_json::Value,
}

/// A CPER-carrying Redfish event delivered by the BMC.
///
/// For small CPERs `cper_data` is populated with the decoded bytes.
/// For large CPERs `additional_data_uri` is set; call
/// [`RasServiceHandle::fetch_cper_data`] to retrieve the binary payload.
#[derive(Debug, Clone)]
pub struct CperEvent {
    pub event_id:            String,
    pub message_id:          String,
    pub severity:            Option<CperSeverity>,
    pub timestamp:           String,
    pub origin_of_condition: Option<String>,
    pub cper_data:           Option<Vec<u8>>,  // inline CPER bytes
    pub additional_data_uri: Option<String>,   // URI for large-CPER retrieval
    pub raw:                 serde_json::Value,
}

impl CperEvent {
    /// Parse a single EventRecord object from a Redfish EventMessage payload.
    pub fn from_event_record(record: &serde_json::Value) -> Self {
        let message_id = record["MessageId"].as_str().unwrap_or("").to_string();
        let severity   = CperSeverity::from_message_id(&message_id);

        // Attempt to decode an inline CPER from known field locations
        let cper_data = ["AdditionalData"]
            .iter()
            .find_map(|&key| {
                record.get(key)?.as_str().and_then(|s| BASE64.decode(s).ok())
            })
            .or_else(|| {
                record.pointer("/Oem/CperData")
                    .and_then(|v| v.as_str())
                    .and_then(|s| BASE64.decode(s).ok())
            });

        CperEvent {
            event_id:            record["EventId"].as_str().unwrap_or("").to_string(),
            message_id,
            severity,
            timestamp:           record["EventTimestamp"].as_str().unwrap_or("").to_string(),
            origin_of_condition: record.pointer("/OriginOfCondition/@odata.id")
                                       .and_then(|v| v.as_str())
                                       .map(|s| s.to_string()),
            cper_data,
            additional_data_uri: record["AdditionalDataURI"].as_str().map(|s| s.to_string()),
            raw:                 record.clone(),
        }
    }
}

/// A Common Platform Action Descriptor (CPAD) to submit to the BMC.
///
/// The BMC uses `partition_id` to route the action to the correct silicon
/// endpoint.  `payload` is the raw binary CPAD blob transmitted base64-encoded
/// in the JSON body.
#[derive(Debug, Clone)]
pub struct CpadRecord {
    pub platform_id:  String,    // identifies this BMC in the fleet
    pub partition_id: String,    // identifies the target silicon endpoint
    pub creator_id:   String,    // identifies the issuing analyzer
    pub payload:      Vec<u8>,   // binary CPAD blob
    pub fru_id:       String,
    pub fru_text:     String,    // physical location label (e.g. silkscreen)
}

// ---------------------------------------------------------------------------
// Service handle
// ---------------------------------------------------------------------------

/// Redfish client handle for the RAS API service.
pub struct RasServiceHandle<'ctx> {
    ctx: &'ctx ClientContext,
}

impl<'ctx> RasServiceHandle<'ctx> {
    pub(crate) fn new(ctx: &'ctx ClientContext) -> Self { Self { ctx } }

    fn service_uri(&self) -> String {
        self.ctx.resolve_service_uri("RasService", "/RasService")
    }

    fn event_subscriptions_uri(&self) -> String {
        let svc = self.ctx.resolve_service_uri("EventService", "/EventService");
        format!("{}/Subscriptions", svc)
    }

    // ── Async ────────────────────────────────────────────────────────────────

    /// Discover all RAS API endpoints exposed by the BMC.
    pub async fn discover_endpoints(&self) -> Result<Vec<RasEndpoint>, RedfishError> {
        let resp = self.ctx.get(&self.service_uri()).await?;
        if !resp.success { return Ok(vec![]); }

        let empty = serde_json::json!([]);
        let members = resp.body
            .as_ref()
            .and_then(|b| b.get("Members"))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let mut endpoints = Vec::new();
        for member in &members {
            let uri = match member.get("@odata.id").and_then(|v| v.as_str()) {
                Some(u) => u,
                None    => continue,
            };
            let detail = self.ctx.get(uri).await?;
            if !detail.success { continue; }
            let null = serde_json::Value::Null;
            let d = detail.body.as_ref().unwrap_or(&null);
            endpoints.push(RasEndpoint {
                endpoint_id:      d["Id"].as_str().unwrap_or("").to_string(),
                creator_id:       d["CreatorId"].as_str().unwrap_or("").to_string(),
                fru_id:           d["FruId"].as_str().unwrap_or("").to_string(),
                partition_id:     d["PartitionId"].as_str().unwrap_or("").to_string(),
                supported_queues: d["SupportedQueues"].as_array()
                    .map(|a| a.iter()
                              .filter_map(|v| v.as_str().map(|s| s.to_string()))
                              .collect())
                    .unwrap_or_default(),
                uri: uri.to_string(),
                raw: d.clone(),
            });
        }
        drop(empty);
        Ok(endpoints)
    }

    /// Subscribe to CPER-carrying Redfish events from this BMC.
    ///
    /// Leave `registry_prefixes` and `message_ids` empty to receive all events
    /// and filter client-side with [`parse_cper_events`].
    pub async fn subscribe_cper_events(
        &self,
        destination:       &str,
        registry_prefixes: Vec<String>,
        message_ids:       Vec<String>,
        context:           &str,
        event_format_type: &str,
    ) -> Result<RedfishResponse, RedfishError> {
        let subs_uri = self.event_subscriptions_uri();
        let mut body = json!({
            "Destination":      destination,
            "Protocol":         "Redfish",
            "SubscriptionType": "RedfishEvent",
            "Context":          context,
            "EventFormatType":  event_format_type,
        });
        if !registry_prefixes.is_empty() {
            body["RegistryPrefixes"] = json!(registry_prefixes);
        }
        if !message_ids.is_empty() {
            body["MessageIds"] = json!(message_ids);
        }
        self.ctx.post(&subs_uri, body).await
    }

    /// Fetch a CPER payload from the BMC via the `AdditionalDataUri` in an event.
    pub async fn fetch_cper_data(&self, additional_data_uri: &str) -> Result<Vec<u8>, RedfishError> {
        let resp = self.ctx.get(additional_data_uri).await?;
        if !resp.success {
            return Err(RedfishError::HttpError {
                status_code: resp.status_code,
                message: format!("Failed to fetch CPER data from {additional_data_uri}"),
            });
        }

        // Try known base64-encoded fields in priority order
        if let Some(body) = &resp.body {
            for key in &["CperData", "Data", "AdditionalData"] {
                if let Some(s) = body.get(key).and_then(|v| v.as_str()) {
                    if let Ok(bytes) = BASE64.decode(s) {
                        return Ok(bytes);
                    }
                }
            }
            // Fallback: return JSON body serialised as bytes
            return Ok(serde_json::to_vec(body).unwrap_or_default());
        }

        Ok(vec![])
    }

    /// Submit a CPAD to the BMC via HTTP PUT.
    ///
    /// Action completion is confirmed asynchronously via a Platform Action Event
    /// CPER on the Event queue after the BMC routes the CPAD to the silicon endpoint.
    pub async fn submit_cpad(
        &self,
        cpad_uri: &str,
        cpad:     &CpadRecord,
    ) -> Result<RedfishResponse, RedfishError> {
        let body = json!({
            "PlatformId":  cpad.platform_id,
            "PartitionId": cpad.partition_id,
            "CreatorId":   cpad.creator_id,
            "FruId":       cpad.fru_id,
            "FruText":     cpad.fru_text,
            "Payload":     BASE64.encode(&cpad.payload),
        });
        self.ctx.put(cpad_uri, body).await
    }

    /// Parse `CperEvent` objects from a Redfish EventMessage payload.
    ///
    /// Non-RAS event records in the same payload are silently ignored.
    pub fn parse_cper_events(event_payload: &serde_json::Value) -> Vec<CperEvent> {
        let events = event_payload.get("Events")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        events.iter()
              .map(CperEvent::from_event_record)
              .filter(|ev| {
                  ev.severity.is_some()
                  || ev.message_id.to_lowercase().contains("cper")
                  || ev.message_id.to_lowercase().contains("ras")
              })
              .collect()
    }

    // ── Blocking wrappers ────────────────────────────────────────────────────

    pub fn discover_endpoints_blocking(&self) -> Result<Vec<RasEndpoint>, RedfishError> {
        self.ctx.runtime.as_ref().expect("no runtime").block_on(self.discover_endpoints())
    }

    pub fn subscribe_cper_events_blocking(
        &self,
        destination:       &str,
        registry_prefixes: Vec<String>,
        message_ids:       Vec<String>,
        context:           &str,
        event_format_type: &str,
    ) -> Result<RedfishResponse, RedfishError> {
        self.ctx.runtime.as_ref().expect("no runtime").block_on(
            self.subscribe_cper_events(destination, registry_prefixes, message_ids, context, event_format_type)
        )
    }

    pub fn fetch_cper_data_blocking(&self, uri: &str) -> Result<Vec<u8>, RedfishError> {
        self.ctx.runtime.as_ref().expect("no runtime").block_on(self.fetch_cper_data(uri))
    }

    pub fn submit_cpad_blocking(&self, cpad_uri: &str, cpad: &CpadRecord) -> Result<RedfishResponse, RedfishError> {
        self.ctx.runtime.as_ref().expect("no runtime").block_on(self.submit_cpad(cpad_uri, cpad))
    }
}
